module/os/linux/spl/spl-kmem-cache.c

```diff
@@ -204,7 +204,12 @@ kv_alloc(spl_kmem_cache_t *skc, int size, int flags)
 		ptr = (void *)__get_free_pages(lflags, get_order(size));
 	} else if (skc->skc_flags & KMC_KVMEM) {
 		ptr = spl_kvmalloc(size, lflags);
-	} else {
+	} else if (skc->skc_obj_align > SPL_KMEM_CACHE_ALIGN) {
```

Comment:
This condition would need to be checked before `skc->skc_flags & KMC_KVMEM` otherwise we might end up with a non-page aligned allocation when a alignment was requested.  If we want to continue to allow `KMC_KVMEM` to at least be attempted we need something more like this.

```c
static void *
kv_alloc(spl_kmem_cache_t *skc, int size, int flags)
{
        gfp_t lflags = kmem_flags_convert(flags);
        void *ptr;

        if (skc->skc_flags & KMC_KMEM) {
                ASSERT(ISP2(size));
                ptr = (void *)__get_free_pages(lflags, get_order(size));
                ASSERT(IS_P2ALIGNED(ptr, PAGE_SIZE));
        } else if ((skc->skc_flags & KMC_KVMEM) && 
            (skc->skc_obj_align <= SPL_KMEM_CACHE_ALIGN)) {
                ptr = spl_kvmalloc(size, lflags);
        } else {
                ptr = __vmalloc(size, lflags | __GFP_HIGHMEM, PAGE_KERNEL);
                ASSERT(IS_P2ALIGNED(ptr, PAGE_SIZE));
        }

        return (ptr);
}
```

Reply:
oh yes, right. thanks!